### DISCUSSION TITLE
Offset out of bounds error when filtering path layer by time stamps.

### DISCUSSION THREAD

**Discussion by weberjavi** [visgl/deck.gl#7825] [Q&A]
Hello, I'm extending a [`PathLayer`](https://github.com/visgl/deck.gl/blob/cc727dc7b7f42b201529d1ebb6ece81a7ce71c9d/modules/layers/src/path-layer/path-layer.ts), and getting some errors when trying to filter the rendered paths by a timestsamp.
This extended layer is then used inside a composite layer

In some cases (haven't found the reason why) the error is triggered and no path is rendered 👇🏼 

```
deck: update of VesselTrackLayer({id: '453dcac16-6c65-22f0-3300-05b62e3f89cb'}): offset is out of bounds
   window.console.error @ next-dev.js:20
   onError @ deck.js:59
   raiseError @ layer.js:516
   _handleError @ layer-manager.js:158
   _updateLayer @ layer-manager.js:262
   _updateSublayersRecursively @ layer-manager.js:213
   _updateLayers @ layer-manager.js:174
   setLayers @ layer-manager.js:143
   updateLayers @ layer-manager.js:151
   _onRenderFrame @ deck.js:720
   onRender @ animation-loop.js:232
   _renderFrame @ animation-loop.js:322
   redraw @ animation-loop.js:157
   renderFrame @ animation-loop.js:256
   requestAnimationFrame (async)
   requestAnimationFrame @ request-animation-frame.js:2
   _requestAnimationFrame @ animation-loop.js:309
   renderFrame @ animation-loop.js:257
```

I've tried to filter the data both through the dataFilter extension and through shaders (the later is the preferred one, since paths are smoothly 'animated').

__with DataFilterExtension__ _inside the composite layer_

```ts
_getVesselTrackLayer() {
    return new VesselTrackLayer<Segment[]>(
      this.getSubLayerProps({
        id: `${this.props.id}`,
        data: `https://gateway.api.globalfishingwatch.org/v2/vessels/...`,
        loaders: [trackLoader],
        widthUnits: 'pixels',
        widthScale: 1,
        wrapLongitude: true,
        jointRounded: true,
        capRounded: true,
        onDataLoad: this.onDataLoad,
        getColor: hexToRgb(this.props.color),
        getWidth: 1,
        startTime: this.props.startTime,
        endTime: this.props.endTime,
        getFilterValue: (d) => d.waypoints.map((p) => p.timestamp),
        filterRange: [this.props.startTime, this.props.endTime],
        extensions: [new DataFilterExtension({ filterSize: 1 })],
      })
    )
  }
```
__with shaders__ _inside VesselTrackLayer that extends PathLayer_

```ts
export class VesselTrackLayer<DataT = any, ExtraProps = {}> extends PathLayer<
  DataT,
  Required<_VesselTrackLayerProps> & ExtraProps
> {
  static layerName = 'VesselTrackLayer'
  static defaultProps = defaultProps
  segments!: Segment[]

  getShaders() {
    const shaders = super.getShaders()
    shaders.inject = {
      'vs:#decl': `\
        attribute float instanceTimestamps;
        attribute float instanceNextTimestamps;
        varying float vTime;
      `,
      // Timestamp of the vertex
      'vs:#main-end': `\
        vTime = instanceTimestamps + (instanceNextTimestamps - instanceTimestamps) * vPathPosition.y / vPathLength;
      `,
      'fs:#decl': `\
        uniform float startTime;
        uniform float endTime;
        varying float vTime;
      `,
      // Drop the segments outside of the time window
      'fs:#main-start': `\
        if(vTime < startTime || vTime > endTime) {
          discard;
        }
      `,
    }
    return shaders
  }

  getSegments() {
    return this.segments
  }

  updateState(params) {
    super.updateState(params)
    this.segments = params.props.data
  }

  initializeState() {
    super.initializeState()
    this.segments = []
    const attributeManager = this.getAttributeManager()
    attributeManager.addInstanced({
      timestamps: {
        size: 1,
        accessor: 'getTimestamps',
        shaderAttributes: {
          instanceTimestamps: {
            vertexOffset: 0,
          },
          instanceNextTimestamps: {
            vertexOffset: 1,
          },
        },
      },
    })
  }

  draw(params) {
    const { startTime, endTime } = this.props

    params.uniforms = {
      ...params.uniforms,
      startTime,
      endTime,
    }

    super.draw(params)
  }
}
```

Both implementations throw errors in the same cases. 
Data loader for the layer return data in this format:
```js
[
   {
       waypoints: [
        {coordinates: [-122.3907988, 37.7664413], timestamp: 1554772579000}
        {coordinates: [-122.3908298,37.7667706], timestamp: 1554772579010}
         ...,
        {coordinates: [-122.4485672, 37.8040182], timestamp: 1554772580200}
       ]
     }
   ]

```
I'm substracting the starting date from the timestamps (to the data timestamps and to the actual time range timestamps)
```ts
export const START_TIMESTAMP = new Date('2012-01-01T00:00:00.000Z').getTime()
```
and also using `Math.fround`

I've  also checked that every filtered coordinates pair has a both `long` and `lat` values (in the right order) and an associated timestamp.

Any hint on where to look at where this error is coming from would be really appreciated.
Thanks

**Comment by weberjavi** [comment]
Is there a way to get more verbose error logs or to debug shaders lyfecicle inject points?

**Comment by ghost** [comment]
I suspect this is an issue in your `trackLoader` or the data you feed to your layer. I don't think this has to do with how you attempt to filter or process the data in the shader.

**Reply by weberjavi** [reply]
Thanks @JannikGM .
I've manged to bypass this error. It seemed that it was not being triggered by some wrong shape of the data I was feeding to the layer but because of some values I set to the `_typedArrayManagerProps` on the `DeckGl` component. Pasting here the explanation I shared on the deckgl slack and the response of @Pessimistress 👇🏼 

_My explanation on the fix_
>I think I found the issue. Setting non default values to` _typedArrayManagerProps` was causing the error. I'd set those to `{overAlloc: 1, poolSize: 0} ` to decrease memory consumption.
Removing this prop on the `DeckGL` component allowed the path to render.
Not sure if this is an expected behaviour derived from the `_typedArrayManagerProps`

_Pessimistress response_
>Invalid getPath are filtered out of the data by the path layer, so they end up having 0 vertices. But getTimestamps does not do any filtering, so you end up trying to assign 1-2 time stamps to a path with 0 length
